File: app/src/main/java/com/pr0gramm/app/services/UserService.java
package com.pr0gramm.app.services;

import android.content.SharedPreferences;
import android.database.sqlite.SQLiteDatabase;
import android.support.annotation.Nullable;

import com.google.common.base.Optional;
import com.google.common.base.Stopwatch;
import com.google.common.collect.FluentIterable;
import com.google.common.collect.ImmutableList;
import com.google.gson.Gson;
import com.pr0gramm.app.Settings;
import com.pr0gramm.app.api.pr0gramm.Api;
import com.pr0gramm.app.api.pr0gramm.LoginCookieHandler;
import com.pr0gramm.app.feed.ContentType;
import com.pr0gramm.app.orm.BenisRecord;
import com.pr0gramm.app.util.BackgroundScheduler;
import com.pr0gramm.app.util.Holder;

import org.immutables.value.Value;
import org.joda.time.Duration;
import org.joda.time.Instant;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.ArrayList;
import java.util.List;
import java.util.Set;
import java.util.concurrent.atomic.AtomicBoolean;

import javax.inject.Inject;
import javax.inject.Singleton;

import rx.Completable;
import rx.Observable;
import rx.functions.Func1;
import rx.subjects.BehaviorSubject;
import rx.subjects.PublishSubject;
import rx.util.async.Async;

import static com.pr0gramm.app.Settings.resetContentTypeSettings;
import static com.pr0gramm.app.util.AndroidUtility.checkNotMainThread;
import static org.joda.time.Duration.standardDays;
import static rx.functions.Actions.empty;

/**
 */
@Singleton
@org.immutables.gson.Gson.TypeAdapters
public class UserService {
    private static final Logger logger = LoggerFactory.getLogger("UserService");

    private static final String KEY_LAST_LOF_OFFSET = "UserService.lastLogLength";
    private static final String KEY_LAST_USER_INFO = "UserService.lastUserInfo";
    private static final String KEY_LAST_LOGIN_STATE = "UserService.lastLoginState";

    private static final LoginState NOT_AUTHORIZED = ImmutableLoginState.builder()
            .id(-1).score(0).mark(0)
            .admin(false).premium(false)
            .authorized(false)
            .build();

    private final Object lock = new Object();

    private final Api api;
    private final VoteService voteService;
    private final SeenService seenService;
    private final InboxService inboxService;
    private final LoginCookieHandler cookieHandler;
    private final SharedPreferences preferences;
    private final Holder<SQLiteDatabase> database;

    private final Gson gson;
    private final Settings settings;

    private final AtomicBoolean fullSyncInProgress = new AtomicBoolean();

    // login state and observable for that.
    private LoginState loginState = NOT_AUTHORIZED;
    private final BehaviorSubject<LoginState> loginStateObservable = BehaviorSubject.create(loginState);

    @Inject
    public UserService(Api api,
                       VoteService voteService,
                       SeenService seenService, InboxService inboxService, LoginCookieHandler cookieHandler,
                       SharedPreferences preferences, Settings settings, Gson gson,
                       SingleShotService sso, Holder<SQLiteDatabase> database) {

        this.api = api;
        this.seenService = seenService;
        this.voteService = voteService;
        this.inboxService = inboxService;
        this.cookieHandler = cookieHandler;
        this.preferences = preferences;
        this.settings = settings;
        this.gson = gson;
        this.database = database;

        // only restore user data if authorized.
        if (cookieHandler.hasCookie()) {
            restoreLatestUserInfo();
        }

        this.cookieHandler.setOnCookieChangedListener(this::onCookieChanged);

        loginStateObservable.subscribe(this::persistLatestLoginState);

        // this is not nice, and will get removed in one or two versions!
        // TODO REMOVE THIS ASAP.
        loginStateObservable
                .filter(state -> state.authorized() && state.uniqueToken() == null)
                .switchMap(state -> api.identifier().subscribeOn(BackgroundScheduler.instance()))
                .map(Api.UserIdentifier::identifier)
                .onErrorResumeNext(Observable.empty())
                .subscribe(this::updateUniqueToken);

        loginStateObservable.subscribe(state -> {
            Track.updateAuthorizedState(state.authorized());
        });
    }

    private LoginState updateLoginState(Func1<ImmutableLoginState, LoginState> transformer) {
        synchronized (lock) {
            LoginState newLoginState = transformer.call(ImmutableLoginState.copyOf(this.loginState));

            // persist and publish
            if (newLoginState != loginState) {
                this.loginState = newLoginState;
                this.loginStateObservable.onNext(newLoginState);
            }

            return newLoginState;
        }
    }

    private LoginState updateLoginStateIfAuthorized(Func1<ImmutableLoginState, LoginState> transformer) {
        return updateLoginState(loginState -> {
            if (loginState.authorized()) {
                return transformer.call(loginState);
            } else {
                return loginState;
            }
        });
    }

    private void updateUniqueToken(@Nullable String uniqueToken) {
        if (uniqueToken == null)
            return;

        updateLoginStateIfAuthorized(loginState -> loginState.withUniqueToken(uniqueToken));
    }

    /**
     * Restore the latest user info from the shared preferences
     */
    private void restoreLatestUserInfo() {
        String lastLoginState = preferences.getString(KEY_LAST_LOGIN_STATE, null);
        String lastUserInfo = preferences.getString(KEY_LAST_USER_INFO, null);

        if (lastLoginState != null) {
            Async.start(() -> gson.fromJson(lastLoginState, LoginState.class), BackgroundScheduler.instance())
                    .onErrorResumeNext(Observable.empty())
                    .doOnNext(info -> logger.info("Restoring login state: {}", info))
                    .map(state -> ImmutableLoginState.copyOf(state).withBenisHistory(loadBenisHistory(state.id())))
                    .subscribe(
                            loginState -> updateLoginState(ignored -> loginState),
                            error -> logger.warn("Could not restore login state: " + error));

        } else if (lastUserInfo != null) {
            // TODO will be removed soon

            Async.start(() -> createLoginState(gson.fromJson(lastUserInfo, Api.Info.class)), BackgroundScheduler.instance())
                    .doOnNext(info -> logger.info("Restoring api user info: {}", info))
                    .filter(info -> cookieHandler.hasCookie())
                    .subscribe(
                            loginState -> updateLoginState(ignored -> loginState),
                            error -> logger.warn("Could not restore user info: " + error));
        }
    }

    private void onCookieChanged() {
        Optional<String> cookie = cookieHandler.getLoginCookie();
        if (!cookie.isPresent()) {
            logout();
        }
    }

    public Observable<LoginProgress> login(String username, String password) {
        return api.login(username, password).flatMap(login -> {
            List<Observable<?>> observables = new ArrayList<>();

            if (login.success()) {
                observables.add(updateCachedUserInfo()
                        .doOnTerminate(() -> updateUniqueToken(login.identifier().orNull()))
                        .toObservable());

                // perform initial sync in background.
                sync().subscribeOn(BackgroundScheduler.instance()).subscribe(
                        empty(),
                        err -> logger.error("Could not perform initial sync during login", err));
            }

            // wait for sync to complete before emitting login result.
            observables.add(Observable.just(new LoginProgress(login)));
            return Observable.concatDelayError(observables).ofType(LoginProgress.class);
        });
    }

    /**
     * Check if we can do authorized requests.
     */
    public boolean isAuthorized() {
        return cookieHandler.hasCookie() && loginState.authorized();
    }

    /**
     * Checks if the user has paid for a pr0mium account
     */
    public boolean isPremiumUser() {
        return cookieHandler.isPaid();
    }

    /**
     * Performs a logout of the user.
     */
    public Completable logout() {
        return Async.<Void>start(() -> {
            updateLoginState(oldState -> NOT_AUTHORIZED);

            // removing cookie from requests
            cookieHandler.clearLoginCookie(false);

            // remove sync id
            preferences.edit()
                    .remove(KEY_LAST_LOF_OFFSET)
                    .remove(KEY_LAST_USER_INFO)
                    .remove(KEY_LAST_LOGIN_STATE)
                    .apply();

            // clear all the vote cache
            voteService.clear();

            // clear the seen items
            seenService.clear();

            // no more read messages.
            inboxService.forgetReadMessage();
            inboxService.publishUnreadMessagesCount(0);

            // and reset the content user, because only signed in users can
            // see the nsfw and nsfl stuff.
            resetContentTypeSettings(settings);

            return null;
        }, BackgroundScheduler.instance()).toCompletable();
    }

    public Observable<LoginState> loginState() {
        return loginStateObservable.asObservable();
    }

    /**
     * Performs a sync. This updates the vote cache with all the votes that
     * where performed since the last call to sync.
     */
    public Observable<Api.Sync> sync() {
        if (!cookieHandler.hasCookie())
            return Observable.empty();

        // tell the sync request where to start
        long lastLogOffset = preferences.getLong(KEY_LAST_LOF_OFFSET, 0L);
        boolean fullSync = lastLogOffset == 0;

        if (fullSync && !fullSyncInProgress.compareAndSet(false, true)) {
            // fail fast if full sync is in already in progress.
            return Observable.empty();
        }

        return api.sync(lastLogOffset).doAfterTerminate(() -> fullSyncInProgress.set(false)).flatMap(response -> {
            inboxService.publishUnreadMessagesCount(response.inboxCount());

            int userId = loginState.id();
            if (userId > 0) {
                // save the current benis value
                BenisRecord.storeValue(database.value(), userId, response.score());

                // and load the current benis history
                Graph scoreGraph = loadBenisHistory(userId);

                updateLoginStateIfAuthorized(loginState -> loginState
                        .withScore(response.score())
                        .withBenisHistory(scoreGraph));
            }

            try {
                voteService.applyVoteActions(response.log());

                // store syncId for next time.
                if (response.logLength() > lastLogOffset) {
                    preferences.edit()
                            .putLong(KEY_LAST_LOF_OFFSET, response.logLength())
                            .apply();
                }
            } catch (Throwable error) {
                return Observable.error(error);
            }

            return Observable.just(response);
        });
    }

    /**
     * Retrieves the user data and stores part of the data in the database.
     */
    public Observable<Api.Info> info(String username) {
        return api.info(username, null);
    }

    /**
     * Retrieves the user data and stores part of the data in the database.
     */
    public Observable<Api.Info> info(String username, Set<ContentType> contentTypes) {
        return api.info(username, ContentType.combine(contentTypes));
    }

    /**
     * Returns information for the current user, if a user is signed in.
     *
     * @return The info, if the user is currently signed in.
     */
    public Observable<Api.Info> info() {
        return Observable.from(getName().asSet()).flatMap(this::info);
    }

    /**
     * Update the cached user info in the background.
     */
    public Completable updateCachedUserInfo() {
        PublishSubject publishSubject = PublishSubject.create();

        info().retry(3)
                .subscribeOn(BackgroundScheduler.instance())
                .doOnNext(info -> {
                    LoginState loginState = createLoginState(info);
                    updateLoginState(old -> loginState);
                })
                .doOnTerminate(publishSubject::onCompleted)
                .subscribe(empty(), error -> logger.warn("Could not update user info.", error));

        return publishSubject.toCompletable();
    }

    /**
     * Persists the given login state to a preference storage.
     */
    private void persistLatestLoginState(LoginState state) {
        try {
            if (state.authorized()) {
                logger.info("persisting logins state now.");

                String encoded = gson.toJson(state);
                preferences.edit()
                        .putString(KEY_LAST_LOGIN_STATE, encoded)
                        .apply();
            } else {
                preferences.edit()
                        .remove(KEY_LAST_LOGIN_STATE)
                        .apply();
            }

        } catch (RuntimeException error) {
            logger.warn("Could not persist latest user info", error);
        }
    }

    private LoginState createLoginState(Api.Info info) {
        checkNotMainThread();

        Api.Info.User user = info.getUser();

        return ImmutableLoginState.builder()
                .authorized(true)
                .id(user.getId())
                .name(user.getName())
                .mark(user.getMark())
                .score(user.getScore())
                .premium(isPremiumUser())
                .admin(userIsAdmin())
                .benisHistory(loadBenisHistory(user.getId()))
                .build();
    }

    public boolean userIsAdmin() {
        return cookieHandler.getCookie().transform(this::isTruthValue).or(false);
    }

    private boolean isTruthValue(LoginCookieHandler.Cookie cookie) {
        String repr = String.valueOf(cookie.admin);
        return repr.startsWith("1") || "true".equalsIgnoreCase(repr);
    }

    private Graph loadBenisHistory(int userId) {
        Stopwatch watch = Stopwatch.createStarted();

        Duration historyLength = standardDays(7);
        Instant now = Instant.now();
        Instant start = now.minus(historyLength);

        // get the values and transform them
        ImmutableList<Graph.Point> points = FluentIterable
                .from(BenisRecord.findValuesLaterThan(database.value(), userId, start))
                .transform(record -> {
                    double x = record.time;
                    return new Graph.Point(x, record.benis);
                }).toList();

        logger.info("Loading benis graph took " + watch);
        return new Graph(start.getMillis(), now.getMillis(), points);
    }

    /**
     * Gets the name of the current user from the cookie. This will only
     * work, if the user is authorized.
     *
     * @return The name of the currently signed in user.
     */
    public Optional<String> getName() {
        return cookieHandler.getCookie().transform(cookie -> cookie.n);
    }


    /**
     * Returns an observable that produces the unique user token, if a hash is currently
     * available. This produces "null", if the user is currently not signed in.
     */
    public Observable<String> userToken() {
        return loginStateObservable.take(1).flatMap(value -> {
            if (value.authorized()) {
                return Observable.just(value.uniqueToken());
            } else {
                return Observable.just(null);
            }
        });
    }

    public Observable<Api.AccountInfo> accountInfo() {
        return api.accountInfo();
    }

    @Value.Immutable
    public interface LoginState {
        @Nullable
        @org.immutables.gson.Gson.Ignore
        Graph benisHistory();

        @Nullable
        String uniqueToken();

        @Nullable
        String name();

        int id();

        int score();

        int mark();

        boolean admin();

        boolean premium();

        boolean authorized();
    }

    public static final class LoginProgress {
        @Nullable
        private final Api.Login login;

        private final float progress;

        LoginProgress(float progress) {
            this.progress = progress;
            this.login = null;
        }

        LoginProgress(@Nullable Api.Login login) {
            this.progress = 1.f;
            this.login = login;
        }

        public float getProgress() {
            return progress;
        }

        public Optional<Api.Login> getLogin() {
            return Optional.fromNullable(login);
        }
    }
}

File: app/src/main/java/com/pr0gramm/app/services/UserService.kt
package com.pr0gramm.app.services

import android.content.SharedPreferences
import androidx.core.content.edit
import com.pr0gramm.app.*
import com.pr0gramm.app.api.pr0gramm.Api
import com.pr0gramm.app.api.pr0gramm.GenericSettings
import com.pr0gramm.app.api.pr0gramm.LoginCookieJar
import com.pr0gramm.app.feed.ContentType
import com.pr0gramm.app.model.config.Config
import com.pr0gramm.app.model.user.LoginCookie
import com.pr0gramm.app.model.user.LoginState
import com.pr0gramm.app.orm.BenisRecord
import com.pr0gramm.app.ui.base.AsyncScope
import com.pr0gramm.app.ui.base.launchIgnoreErrors
import com.pr0gramm.app.util.*
import com.squareup.moshi.adapter
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.NonCancellable
import kotlinx.coroutines.flow.*
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import okhttp3.Cookie
import java.util.*
import java.util.concurrent.atomic.AtomicBoolean
import java.util.concurrent.locks.ReentrantLock
import kotlin.concurrent.withLock


class UserService(
    private val api: Api,
    private val voteService: VoteService,
    private val seenService: SeenService,
    private val inboxService: InboxService,
    private val cookieJar: LoginCookieJar,
    private val preferences: SharedPreferences,
    private val benisService: BenisRecordService,
    private val collectionsItemsService: CollectionItemsService,
    private val favedCommentService: FavedCommentService,
    private val settingsService: SiteSettingsService,
    private val config: Config,
) {

    private val logger = Logger("UserService")

    @Suppress("PrivatePropertyName")
    private val NotAuthorized = LoginState(
        id = -1, score = 0, mark = 0, admin = false,
        premium = false, authorized = false, name = null,
        uniqueToken = null, verified = false,
    )

    private val fullSyncInProgress = AtomicBoolean()

    private val loginStateLock = ReentrantLock()
    private val loginStateSubject = MutableStateFlow(NotAuthorized)

    val loginState: LoginState get() = loginStateSubject.value
    val loginStates: Flow<LoginState> = loginStateSubject

    init {
        restoreLatestUserInfo()

        doInBackground {
            // observe changes to the login cookie to handle login/logout behaviour.
            cookieJar.observeCookie.distinctUntilChanged().collect { onCookieChanged(it) }
        }

        AsyncScope.launch {
            // persist the login state every time it changes.
            loginStates.collect { state -> persistLatestLoginState(state) }
        }
    }

    private fun updateLoginState(state: LoginState) {
        if (this.loginStateSubject.value == state) {
            return
        }

        logger.debug { "LoginState is now: $state" }

        // update the current 'auth' state for tracking
        Track.updateUserState(Track.AuthState(state.authorized, state.premium))

        this.loginStateSubject.value = state
    }


    private fun updateUniqueTokenIfNeeded() {
        val state = loginState
        if (state.authorized && state.uniqueToken == null) {
            doInBackground {
                val result = api.identifier()
                updateUniqueToken(result.identifier)
            }
        }
    }

    private fun updateUniqueToken(uniqueToken: String?) {
        if (uniqueToken == null)
            return

        loginStateLock.withLock {
            val updatedLoginState = loginState.copy(uniqueToken = uniqueToken)
            updateLoginState(updatedLoginState)
        }
    }

    /**
     * Restore the latest user info from the shared preferences
     */
    private fun restoreLatestUserInfo() {
        val encodedLoginState = preferences.getString(KEY_LAST_LOGIN_STATE, null)
        logger.debug { "Found login state: $encodedLoginState" }

        if (encodedLoginState == null || encodedLoginState == "null") {
            if (cookieJar.hasCookie()) {
                logger.warn { "Got cookie but no loginState, discarding cookie." }
                cookieJar.clearLoginCookie()
            }

            return
        }

        if (!cookieJar.hasCookie()) {
            logger.warn { "Got loginState but no cookie, discarding login state." }
            preferences.edit { remove(KEY_LAST_LOGIN_STATE) }
            return
        }

        doInBackground {
            try {
                val loginState = MoshiInstance.adapter<LoginState>()
                    .fromJson(encodedLoginState) ?: return@doInBackground

                logger.debug { "Restoring login state: $loginState" }
                updateLoginState(loginState)

                updateUniqueTokenIfNeeded()

            } catch (err: Exception) {
                logger.warn("Could not restore login state:", err)
                logout()
            }
        }
    }

    private fun onCookieChanged(cookie: LoginCookie?) {
        if (cookie == null) {
            logger.info { "LoginCookie was removed, performing logout now." }
            AsyncScope.launch { logout() }

        } else {
            // cookie has changed, re-build cached user info
            AsyncScope.launchIgnoreErrors {
                updateCachedUserInfo()
            }
        }
    }

    suspend fun login(username: String, password: String, captchaToken: String, captchaAnswer: String): LoginResult {
        logger.debug {
            // 'debug' is not compiled into release builds, this is only visible during development!
            "Login with $username/$password (token=$captchaToken, captcha=$captchaAnswer)"
        }

        val response = api.login(username, password, captchaToken, captchaAnswer)

        // in case of errors, just return the Failure
        val login = response.body() ?: run {
            logger.debug { "Request failed, no body." }
            return LoginResult.FailureError
        }

        login.banInfo?.let { banInfo ->
            logger.debug { "User is banned." }
            return LoginResult.Banned(banInfo)
        }

        // handle error codes
        when (login.error) {
            "invalidLogin" -> return LoginResult.FailureLogin
            "invalidCaptcha" -> return LoginResult.FailureCaptcha
        }

        // probably redundant.
        if (login.success == false) {
            logger.debug { "Field login.success is false" }
            return LoginResult.FailureLogin
        }

        // extract login cookie from response
        val loginCookie = Cookie
            .parseAll(response.raw().request.url, response.raw().headers)
            .firstOrNull { cookie -> cookie.name == "me" }

        if (loginCookie == null) {
            logger.debug { "No login cookie found" }
            return LoginResult.FailureError
        }

        // store the cookie
        if (!cookieJar.updateLoginCookie(loginCookie)) {
            logger.debug { "CookieJar did not accept cookie $loginCookie" }
            return LoginResult.FailureError
        }

        // update app settings
        login.settings?.let { settings ->
            settingsService.updateLocalSettings(settings)
        }

        val userInfo = updateCachedUserInfo(login.identifier)
        if (userInfo == null) {
            catchAll { logout() }
            throw IllegalStateException("Could not fetch initial user info")
        }

        // start sync now
        doInBackground { sync() }

        return LoginResult.Success
    }

    /**
     * Performs a logout of the user.
     */
    suspend fun logout() = withContext(NonCancellable + Dispatchers.Default) {
        updateLoginState(NotAuthorized)

        // removing cookie from requests
        cookieJar.clearLoginCookie()

        // remove sync id
        preferences.edit {
            preferences.all.keys
                .filter { it.startsWith(KEY_LAST_LOF_OFFSET) }
                .forEach { remove(it) }

            remove(KEY_LAST_USER_INFO)
            remove(KEY_LAST_LOGIN_STATE)
        }

        // clear all the vote cache
        voteService.clear()

        // clear the seen items
        seenService.clear()

        // remove all cached collection items
        collectionsItemsService.clear()

        // remove all faved comments
        favedCommentService.clear()

        // no more read messages.
        inboxService.forgetUnreadMessages()
        inboxService.publishUnreadMessagesCount(Api.InboxCounts())

        // and reset the content user, because only signed in users can
        // see the nsfw and nsfl stuff.
        Settings.resetContentTypeSettings()

        // do not load automatically anymore
        Settings.feedStartWithUri = null
    }

    val selectedContentType: Set<ContentType>
        get() {
            return if (isAuthorized) {
                Settings.contentType
            } else {
                EnumSet.of(ContentType.SFW)
            }
        }

    val selectedContentTypes: Flow<Set<ContentType>> = run {
        val types = loginStates.flatMapLatest { loginState ->
            if (loginState.authorized) {
                Settings.changes()
                    .map { Settings.contentType }
                    .onStart { emit(Settings.contentType) }
            } else {
                flowOf(EnumSet.of(ContentType.SFW))
            }
        }

        types.distinctUntilChanged()
    }

    /**
     * Performs a sync. This updates the vote cache with all the votes that
     * where performed since the last call to sync.
     */
    suspend fun sync(): Api.Sync? {
        if (!cookieJar.hasCookie()) {
            return null
        }

        // tell the sync request where to start
        val lastLogOffset = preferences.getLong(syncOffsetKey(), 0L)
        val fullSync = (lastLogOffset == 0L)

        if (fullSync && !fullSyncInProgress.compareAndSet(false, true)) {
            // fail if full sync is in already in progress.
            return null
        }

        try {
            val response = api.sync(lastLogOffset)

            // update app settings if changed
            response.settings?.let { settings ->
                settingsService.updateLocalSettings(settings)
            }

            inboxService.publishUnreadMessagesCount(response.inbox)

            val userId = loginState.id
            if (userId > 0) {
                // save the current benis value
                benisService.storeValue(userId, response.score)

                loginStateLock.withLock {
                    // and update login state
                    val updatedLoginState = loginState.copy(score = response.score)
                    if (updatedLoginState.authorized) {
                        updateLoginState(updatedLoginState)
                    }
                }
            }

            voteService.applyVoteActions(response.log)

            // store syncId for next time.
            if (response.logLength > lastLogOffset) {
                preferences.edit {
                    putLong(syncOffsetKey(), response.logLength)
                }
            }

            debugOnly {
                debugConfig.pendingNotifications?.let { pending ->
                    val counts = Api.InboxCounts(comments = pending.messages.size)
                    return Api.Sync(response.logLength, response.log, response.score, counts)
                }
            }

            return response

        } finally {
            if (fullSync) {
                fullSyncInProgress.set(false)
            }
        }
    }

    private suspend fun querySyncOnly(): Api.Sync? {
        if (!cookieJar.hasCookie()) {
            return null
        }

        val lastLogOffset = preferences.getLong(syncOffsetKey(), 0L)
        return api.sync(lastLogOffset)
    }

    private fun syncOffsetKey(): String {
        return KEY_LAST_LOF_OFFSET + ":v6:" + config.syncVersion
    }

    /**
     * Retrieves the user data and stores part of the data in the database.
     */
    suspend fun info(username: String): Api.Info {
        return api.info(username, null)
    }

    /**
     * Retrieves the user data and stores part of the data in the database.
     */
    suspend fun info(username: String, contentTypes: Set<ContentType>): Api.Info {
        return api.info(username, ContentType.combine(contentTypes))
    }

    /**
     * Store the given settings
     */
    suspend fun storeSiteSettings(update: Map<String, Any?>) {
        val sync = runCatching { querySyncOnly() }.getOrNull() ?: return

        val settings = HashMap(sync.settings.orEmpty() + update)

        logger.info { "Storing remote settings: $settings" }

        val fields = settings.mapValues { (_, value) ->
            when (value) {
                is Boolean -> value.toInt()
                else -> value
            }
        }

        return api.storeSettings(null, fields as GenericSettings)
    }

    /**
     * Update the cached user info in the background. Will throw an exception if
     * the user info can not be updated.
     */
    suspend fun updateCachedUserInfo(token: String? = null): Api.Info? {
        val name = name.takeUnless { it.isNullOrEmpty() } ?: return null

        return info(name).also { userInfo ->
            val updatedLoginState = createLoginStateFromInfo(
                userInfo.user,
                cookieJar.parsedCookie, token ?: loginState.uniqueToken
            )

            loginStateLock.withLock {
                if (cookieJar.parsedCookie != null) {
                    updateLoginState(updatedLoginState)
                }
            }
        }
    }

    /**
     * Persists the given login state to a preference storage.
     */
    private fun persistLatestLoginState(state: LoginState) {
        try {
            if (state.authorized) {
                logger.debug { "Persisting login state now: $loginState" }

                preferences.edit {
                    val encoded = MoshiInstance.adapter<LoginState>().toJson(state)
                    putString(KEY_LAST_LOGIN_STATE, encoded)
                }
            } else {
                preferences.edit {
                    remove(KEY_LAST_LOGIN_STATE)
                }
            }

        } catch (error: RuntimeException) {
            logger.warn("Could not persist latest user info", error)
        }

    }

    val isAuthorized: Boolean
        get() = loginState.authorized && cookieJar.parsedCookie?.id?.isNotBlank() == true

    val userIsPremium: Boolean
        get() = isAuthorized && loginState.premium

    val userIsAdmin: Boolean
        get() = isAuthorized && loginState.admin

    val userIsVerified: Boolean
        get() = isAuthorized && loginState.verified

    val loginStateWithBenisGraph = run {
        val graphs = loginStates.mapLatest { loginState ->
            if (loginState.authorized) {
                LoginStateWithScoreGraph(loginState, loadScoreHistoryAsGraph(loginState.id))
            } else {
                LoginStateWithScoreGraph(loginState, scoreGraph = null)
            }
        }

        ticker(Duration.minutes(10)).flatMapLatest { graphs }
    }

    private suspend fun loadScoreHistoryAsGraph(userId: Int): Graph = logger.time("Loading benis graph") {
        val now = Instant.now()
        val start = now - Duration.days(7)

        // get the values and transform them
        val points = benisService.findValuesLaterThan(userId, start).map { record ->
            val x = record.time.toDouble()
            val y = record.benis.toDouble()
            Graph.Point(x, y)
        }

        return Graph(start.millis.toDouble(), now.millis.toDouble(), optimizeValuesBy(points) { it.y })
    }

    /**
     * Loads all benis records for the current user.
     */
    suspend fun loadBenisRecords(after: Instant = Instant(0)): BenisGraphRecords {
        return BenisGraphRecords(benisService.findValuesLaterThan(loginState.id, after))
    }

    class BenisGraphRecords(val records: List<BenisRecord>)

    /**
     * Gets the name of the current user from the cookie. This will only
     * work, if the user is authorized.

     * @return The name of the currently signed in user.
     */
    val name: String?
        get() = cookieJar.parsedCookie?.name

    suspend fun requestPasswordRecovery(email: String) {
        api.requestPasswordRecovery(email)
    }

    suspend fun resetPassword(name: String, token: String, password: String): Boolean {
        val result = api.resetPassword(name, token, password)
        return result.error == null
    }

    suspend fun userCaptcha(): Api.UserCaptcha {
        return api.userCaptcha()
    }

    suspend fun blockUser(name: String) {
        api.block(null, name)
    }

    class LoginStateWithScoreGraph(
        val loginState: LoginState, val scoreGraph: Graph?
    )

    sealed class LoginResult {
        object Success : LoginResult()

        object FailureError : LoginResult()
        object FailureCaptcha : LoginResult()
        object FailureLogin : LoginResult()

        class Banned(val ban: Api.Login.BanInfo) : LoginResult()
    }
}

private const val KEY_LAST_LOF_OFFSET = "UserService.lastLogLength"
private const val KEY_LAST_USER_INFO = "UserService.lastUserInfo"
private const val KEY_LAST_LOGIN_STATE = "UserService.lastLoginState"

private fun createLoginStateFromInfo(user: Api.Info.User, cookie: LoginCookie?, token: String?): LoginState {
    return LoginState(
        authorized = true,
        id = user.id,
        name = user.name,
        mark = user.mark,
        score = user.score,
        premium = cookie?.paid == true,
        admin = cookie?.admin == true,
        verified = cookie?.verified == true,
        uniqueToken = token
    )
}